Make codegen backend agnostic minus fallbacks

bdhirsh opened the pull request:
This PR removes all xla-specific logic from the pytorch codegen, except for the CPU fallback logic (working on in a later PR [here](https://github.com/pytorch/xla/pull/2945). The corresponding pytorch-side changes are https://github.com/pytorch/pytorch/pull/58064 and https://github.com/pytorch/pytorch/pull/58568.

Main changes:
- Rename `aten_xla_type.h` -> `XLANativeFunctions.h`, which is a little more in-line with how we name other in-tree generated files. Also updated the corresponding documentation.
- I deleted `InitializeAtenBindings()` mostly because it looks like legacy (it's a no-op), and it would be an extra hoop to force every external backend to implement the same function
- Removed the `AtenXlaType` class. XLA lowering are now just defined under `torch_xla::{op}`.
- I also started passing a few xla macros and headers directly through the yaml, instead of hardcoding them in the codegen. I'm actually hoping to remove them though, since the logging information seems redundant: they're used in the generate `out`/`inplace` wrappers, but those wrapper just call into the handwritten xla functional lowering, which also has its own logging information.

bdhirsh commented on xla_native_functions.yaml:
```diff
@@ -52,6 +61,8 @@ supported:
   - convolution_overrideable
   - convolution_backward_overrideable
   - _copy_from
+  - _copy_from_and_resize
```
Huh, not actually sure why these are showing up in the diff, since they're already there in master.

bdhirsh replied:
fixed

bdhirsh commented on xla_native_functions.yaml:
```diff
@@ -1,5 +1,15 @@
 backend: XLA
 cpp_namespace: torch_xla
+per_op_log: XLA_FN_TRACK(3)
+per_argument_log: TF_VLOG(3)
```
@ailzhang @JackCaoG I'm actually hoping that I can just completely remove this, but I wanted to confirm with you. Right now, the auto-generated `add_out()` kernel for XLA has the same logging behavior as the original XLA codegen, and looks like this:
```
at::Tensor wrapper_Tensor_add(const at::Tensor &self, const at::Tensor &other,
                              const at::Scalar &alpha) {
  // This is the actual XLA add kernel
  return torch_xla::add(self, other, alpha);
}

}  // anonymous namespace
at::Tensor &wrapper_out_add_out(const at::Tensor &self, const at::Tensor &other,
                                const at::Scalar &alpha, at::Tensor &out) {
  XLA_FN_TRACK(3);
  TF_VLOG(3) << "XLA wrapper_out_add_out :"
             << " self=" << self.toString() << " other=" << other.toString()
             << " out=" << out.toString();

  auto wrapper_out_add_out_tmp = wrapper_Tensor_add(self, other, alpha);
  at::_copy_from_and_resize(wrapper_out_add_out_tmp, out);
  return out;
}
```

That logging information in the generated `add_out` kernel seems unnecessary though, since it ends up calling directly into `torch_xla::add`, which [has its own logging too.](https://github.com/pytorch/xla/blob/280348439b23fd7aef5c57c9b92298f8890bcc98/torch_xla/csrc/aten_xla_type.cpp#L501).

Passing C++ macros directly into yaml so the codegen can plop them into a C++ function is also pretty fragile; what if another backend wants some other more custom logging.

So I think it would be cleaner to kill that code-generated logging and just expect all logging to be done in the functional kernel lowerings, but I wanted to see what your thoughts are!

JackCaoG replied:
I agree that `auto-generated add_out()` should not have logging and counter if the actual kernel has the logging and counter already.

JackCaoG commented on OP_LOWERING_GUIDE.md:
```diff
@@ -12,10 +12,11 @@ You can find the definition of the C++ ATen operations in [native_functions.yaml
-2. `aten_xla_type.h/.cpp` are entry points of PyTorch to the pytorch_xla world. `aten_xla_type.h` is auto-generated through a combination of `xla_native_functions.yaml` and the PyTorch core `native_functions.yaml` file, and contains declarations for kernels that need to be defined in `aten_xla_type.cpp`. The kernels written here need to construct 'XLATensor' using the input `at::Tensor` and other ...
+1. `xla_native_functions.yaml` contains the list of all operators that are lowered. Each operator name must directly match a pytorch operator listed in [native_functions.yaml](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml). This file serves as the interface to adding new xla operators, and is an input to PyTorch's [codegen machinery](https://github.com/ ...
+2. `XLANativeFunctions.h` and `aten_xla_type.cpp` are entry points of PyTorch to the pytorch_xla world, and contain the manually written lowerings to XLA for each operator. `XLANativeFunctions.h` is auto-generated through a combination of `xla_native_functions.yaml` and the PyTorch core `native_functions.yaml` file, and contains declarations for kernels that need to be defined in `aten_xla_type.c ...
+2. `RegisterXLA.cpp` and `RegisterAutogradXLA.cpp` are auto-generated files that register all lowerings to the PyTorch Dispatcher. They also include auto-generated wrapper implementations of `out=` and `inplace` operators.
```
Could you change `2.` to `3.` and update the order in below.

JackCaoG commented on OP_LOWERING_GUIDE.md:
```diff
@@ -12,10 +12,11 @@ You can find the definition of the C++ ATen operations in [native_functions.yaml
+2. `RegisterXLA.cpp` and `RegisterAutogradXLA.cpp` are auto-generated files that register all lowerings to the PyTorch Dispatcher. They also include auto-generated wrapper implementations of `out=` and `inplace` operators.
 3. `aten_xla_type_default.h/.cpp` are also auto-generated, and contain our default implementation of the PyTorch operations which simply fall back to the underlying CPU implementation. Functions in here will be used if lowering is not explicitly defined in `xla_native_functions.yaml` + `aten_xla_type.cpp`.
-4. `tensor.h` contains the `XLATensor` declarations. These declarations are one to one mapping of the `at::Tensor` nodes we declared in `aten_xla_type.h`
+4. `tensor.h` contains the `XLATensor` declarations. These declarations are one to one mapping of the `at::Tensor` nodes we declared in `XLANativeFunctions.h`
```
Not directly related to your change but could you update "These declarations are one to one mapping" to something similar to "These declarations are usually one to one mapping". We sometimes reuse `XLATensor` for multiple atenXlaType functions.

JackCaoG reviewed (changes requested):
Mostly LGTM, some minor nits.

bdhirsh commented:
Fixed the docs. I also removed the logging from the generated inplace/out wrappers, which simplifies some of the codegen and makes the yaml more robust.

JackCaoG reviewed (approved):
Thanks @bdhirsh !
